File: devloop/history.py
from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path


def history_path(checkpoint_path) -> Path:
    """history.jsonl 與 checkpoint 同目錄。"""
    return Path(checkpoint_path).parent / "history.jsonl"


def append_history(checkpoint_path, event, from_phase, to_phase, iteration) -> None:
    """對 history.jsonl 追加一筆 transition 紀錄(append-only,一行一 JSON)。"""
    entry = {
        "ts": datetime.now(timezone.utc).isoformat(),
        "event": event,
        "from": from_phase,
        "to": to_phase,
        "iteration": iteration,
    }
    path = history_path(checkpoint_path)
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("a", encoding="utf-8") as fh:
        fh.write(json.dumps(entry, ensure_ascii=False) + "\n")

File: plugins/dev-loop/src/history.ts
import { writeFileSync, mkdirSync } from "node:fs";
import { dirname, resolve } from "node:path";

/**
 * Computes the path to the history.jsonl file alongside a checkpoint.
 */
export function historyPath(checkpointPath: string): string {
  const checkpointDir = dirname(checkpointPath);
  return resolve(checkpointDir, "history.jsonl");
}

/**
 * Appends a state transition record to the history file (append-only, one JSON object per line).
 * Creates the parent directory if it does not exist.
 */
export function appendHistory(
  checkpointPath: string,
  event: string,
  fromPhase: string,
  toPhase: string,
  iteration: number,
): void {
  const entry = {
    ts: new Date().toISOString(),
    event,
    from: fromPhase,
    to: toPhase,
    iteration,
  };

  const histPath = historyPath(checkpointPath);
  const histDir = dirname(histPath);
  mkdirSync(histDir, { recursive: true });

  writeFileSync(histPath, JSON.stringify(entry) + "\n", { flag: "a", encoding: "utf-8" });
}
